File: linked_lists/fast_slow_pointers/c++/delete_k_node/delete_k_node.cpp
// delete_k_node.cpp
#include "delete_k_node.hpp"
#include "node.hpp"

ListNode::Ptr deleteKthNode(const ListNode::Ptr &head, int k)
{
    // Single node after sentinel head
    if (head->next->next == nullptr)
    {
        // Remove the single node as it is considered the middle node
        head->next = nullptr;
        return head;
    }

    // Shared pointers instead of raw pointers
    ListNode::Ptr fast = head;
    ListNode::Ptr slow = head;
    ListNode::Ptr prev = head;

    // Move the fast pointer k steps ahead of slow
    for (int i = 0; i < k; i++)
    {
        fast = fast->next;
    }

    // Move fast and slow pointers at the same speed and use prev to track the (k - 1)th node
    while (fast != nullptr)
    {
        fast = fast->next;
        // Set prev to slow before iterating slow to the next node
        prev = slow;
        slow = slow->next;
    }

    // Remove the node pointed to by the slow pointer
    prev->next = slow->next;

    return head;
}

File: linked_lists/fast_slow_pointers/python3/delete_k_node.py
from node import ListNode
from singly import SinglyLinkedList


def delete_k_node(head: ListNode, k: int) -> ListNode:
    """
    Given the `head` of a singly linked list, remove the `kth` node from the end of the list.

    Parameters
    ----------
    head : ListNode
        The sentinel head of the linked list
    k : int
        The node to remove

    Returns
    -------
    ListNode
        The sentinel head of the modified linked list with the kth node from the end removed
    """
    # If the real node after the sentinel head is the only node
    if head.next.next == None:
        # Remove it since it is the only node to remove, assuming k is correctly specified as 1
        head.next = None
        return head 
    
    # Pointers initialized to point to sentinel head
    fast, slow, prev = head, head, head
    
    # Move fast k steps ahead of slow
    for _ in range(k):
        fast = fast.next
    
    # Move fast and slow at the same speed and use prev to track the (k - 1)th node
    while fast:
        fast = fast.next
        # Set previous to slow before iterating slow 
        prev = slow
        slow = slow.next

    # Remove the kth node from the end
    prev.next = slow.next

    return head

def main() -> int:

    test_cases = [([1, 2], 2), ([1, 2], 1), ([1, 2, 3, 4, 5], 2), ([1], 1), ([2, 7, 12, 5, 9, 8, 13, 17], 5)]
    for data_array, k in test_cases:
        nodes = (ListNode(data=data) for data in data_array)
        sll = SinglyLinkedList()
        for node in nodes:
            sll.add_to_end(node)

        print(f"Before removing the k = {k} node from the end")
        sll.display()
        head = delete_k_node(sll.head, k)
        print(f"After removing the k = {k} node from the end")
        sll.display()
        print('\n')

    return 0

if __name__ == "__main__":

    main()
        
